imap/relocate_by_id.c

```diff
@@ -255,9 +258,46 @@ int main(int argc, char **argv)
                 }
 
                 /* Add xapian tier paths */
-                struct search_rock rock =
-                    { argv[i], userpath, oldpaths, newpaths };
-                config_foreachoverflowstring(get_searchparts, &rock);
+                char *activefname = user_hash_meta(userid, FNAME_XAPIANSUFFIX);
+                struct mappedfile *activefile = NULL;
+
+                r = mappedfile_open(&activefile, activefname, 0);
+                if (r) {
+                    fprintf(stderr,
+                            "Failed to open activefile for %s: %s\n",
+                            activefname, error_message(r));
+                    goto cleanup;
+                }
+
+                strarray_t *items = strarray_nsplit(mappedfile_base(activefile),
+                        mappedfile_size(activefile), NULL, 1);
+
+                struct buf buf = BUF_INITIALIZER;
+                if (items) {
+                    struct search_rock rock = {
+                        userid, userpath,
+                        STRARRAY_INITIALIZER,
+                        ARRAYU64_INITIALIZER,
+                        oldpaths, newpaths,
+                    };
+                    int j;
+                    for (j = 0; j < strarray_size(items); j++) {
+                        const char *item = strarray_nth(items, j);
+                        const char *col = strrchr(item, ':');
+                        if (!col) continue;
+                        buf_setmap(&buf, item, col-item);
+                        strarray_append(&rock.tiernames, buf_cstring(&buf));
+                        arrayu64_append(&rock.tiergens, atoi(col+1));
+                    }
+                    config_foreachoverflowstring(get_searchparts, &rock);
```

Comment:
I suspect I would have done this the other way out - just resolved the on-disk path for each `tier:gen` with old-style and new-style names and renamed those (if old existed and new didn't) rather than using overflowstrings at all.

Reply:
True, but traversing the config saves us figuring out the partition names and their respective searchpartition config keys. Of course if you want, we can change the current scheme.